opds_import.py

```diff
@@ -288,53 +360,148 @@ def import_from_metadata(
+        fp_data_meta, fp_data_circ, status_messages, next_links = self.extract_data_from_feedparser(feed=feed, data_source=self.data_source_name)
+        # gets: medium, measurements, links, contributors, etc.
+        xml_data_meta = self.extract_metadata_from_elementtree(feed)
+
+        # translate the id in status_messages to identifier.urn
+        identified_messages = {}
+        for id, message_data in status_messages.items():
+            external_identifier, ignore = Identifier.parse_urn(self._db, id)
+            if self.identifier_mapping:
+                internal_identifier = self.identifier_mapping.get(
+                    external_identifier, external_identifier)
+            else:
+                internal_identifier = external_identifier
+            identified_messages[internal_identifier.urn] = message_data
 
+        # Use one loop for both, since the id will be the same for both dictionaries.
+        metadata = {}
+        circulationdata = {}
+        for id, m_data_dict in fp_data_meta.items():
             external_identifier, ignore = Identifier.parse_urn(self._db, id)
             if self.identifier_mapping:
                 internal_identifier = self.identifier_mapping.get(
                     external_identifier, external_identifier)
             else:
                 internal_identifier = external_identifier
-            combined['primary_identifier'] = IdentifierData(
+
+            identifier_obj = IdentifierData(
                 type=internal_identifier.type,
                 identifier=internal_identifier.identifier
             )
-            metadata.append(Metadata(**combined))
-        return metadata, status_messages, next_links
+
+            # form the Metadata object
+            xml_data_dict = xml_data_meta.get(id, {})
+            combined_meta = self.combine(m_data_dict, xml_data_dict)
+            if combined_meta.get('data_source') is None:
+                combined_meta['data_source'] = self.data_source_name
+            
+            combined_meta['primary_identifier'] = identifier_obj
+
+            metadata[internal_identifier.urn] = Metadata(**combined_meta)
+
+            # form the CirculationData that would correspond to this Metadata
+            c_data_dict = fp_data_circ[id]
+
+            circ_links_dict = {}
+            # extract just the links to pass to CirculationData constructor
+            circ_links_dict['links'] = xml_data_dict['links']
+            combined_circ = self.combine(c_data_dict, circ_links_dict)
+            if combined_circ.get('data_source') is None:
+                combined_circ['data_source'] = self.data_source_name
+            
+            combined_circ['primary_identifier'] = identifier_obj
+
+            circulationdata[internal_identifier.urn] = CirculationData(**combined_circ)
+            if id in fp_data_circ:
+                del fp_data_circ[id]
+
+        for id, fp_data_dict in fp_data_circ.items():
```

Comment:
if you put the circulation data in the metadata, you shouldn't need to maintain fp_data_circ and you can take out this part